hydro_lang/src/cycle.rs

```diff
@@ -28,45 +28,63 @@ pub trait CycleCollectionWithInitial<'a, T>: CycleComplete<'a, T> {
 /// by a stream that is not yet known.
 ///
 /// See [`crate::FlowBuilder`] for an explainer on the type parameters.
-pub struct ForwardRef<'a, S: CycleComplete<'a, ForwardRefMarker>> {
+pub struct ForwardRef<'a, S>
```

Comment:
I'm a bit less sure on moving constraints to `where` clauses for structs / impls, it makes the signature of the structure a bit more confusing since it "appears" to allow for any type. For functions though the changes seem totally good.

@MingweiSamuel thoughts?

Reply:
Okay! This seems reasonable, let's go with `where` clauses. Yeah for `impl`s I also prefer having short single-letter names.

Reply:
I've been referring to that type parameter as `Tag` in docs and elsewhere, so maybe that? `S` there refers to the type of "Stream" that will be completed (`Stream` or `Singleton` or `Optional`).